[refs] `lib.convert` preserves `:ident`s through a round trip

bshepherdson opened the pull request:
**NOTE:** This PR got reverted by #51128 and the second edition is #51133.


Closes #49682. Closes #49683.

### Description

Updates `lib.convert` and legacy MBQL normalization to allow
`:ident`s, and fixes some QP middleware and utils so they can
still match up refs correctly.

No user-visible impact.

Newly created queries will have `:ident`s generated for joins,
aggregations, etc. by the previous query, and this will start
saving them into appDBs.

Pre-existing clauses don't have refs added yet.

### How to verify

Newly created "columnar" clauses (joins, aggregations, breakouts, expressions)
get `:ident`s generated by the lib, and these are now visible in legacy MBQL
and on the wire.

If you save a question, the generated `:ident`s will be returned when we read the
card again.

See the unit test changes for lots of examples of what this changed in legacy MBQL.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

bshepherdson commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/50085?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#50085** <a href="https://app.graphite.dev/github/pr/metabase/metabase/50085?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/metabase/metabase/50085?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#50083** <a href="https://app.graphite.dev/github/pr/metabase/metabase/50083?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

bshepherdson committed:
[refs] `lib.convert` preserves `:ident`s through a round trip

Also updates legacy MBQL normalization to allow `:ident`s, and fixes
some QP middleware and utils so they can still match up refs.

No user-visible impact.

Newly created queries will have `:ident`s generated for joins,
aggregations, etc. by the previous query, and this will start
saving them into appDBs.

Pre-existing clauses don't have refs added yet.

appleby reviewed (commented):
If I check out this branch and build a query in the gui then click "visualize" I see the following in my API request data to the `dataset` endpoint (and my join clause doesn't have an `:ident`)

```json
"aggregation-idents": {
    "0": null
},
"breakout-idents": {
    "0": null
},
```

if I then save the query the idents are neither sent over the wire nor persisted to the db.

Is this expected or do I need to test in the branch that corresponds to the latest PR in the stack? Also possible my build is in a weird state. Let me know if this doesn't seem right I can dig further.

For reference the example query I'm using is

- Orders
- explicit join Products on Product ID
- filter: Total > 100
- summarize: count, by: Products -> Category

bshepherdson commented on test/metabase/lib/convert_test.cljc:
```diff
@@ -427,7 +386,8 @@
      :query    {:source-table 224
-                :expressions {"a" 1}}}
+                :expressions {"a" 1}
+                :expression-idents {"a" (u/generate-nano-id)}}}
```
They're opaque, random strings. Most tests should use `mt/mbql-query` and friends, which will (1) automatically generate `:expression-idents` etc.; and (2) has metadata that makes `=?` ignore the generated idents on the `expected` side.

bshepherdson replied:
To amplify that first statement: these are opaque, unique strings. We currently happen to be using NanoIDs when we need a new `:ident` _ex nihilo_ for a newly created aggregation/breakout/expression/join, but nothing should be **reading** these idents.

The only legal operations on these idents are:
- matching by `=` (including as map keys), and
- concatenating other strings to the ends
    - eg. to attach a fragment for a join to each joined column's ident

appleby replied:
@bshepherdson my comment was badly worded, but by "painful" I was thinking of the potential need to migrate stored ids if we changed the format (but maybe even that is not necessary).

Really I just meant "unlikely" rather than "painful"

bshepherdson committed:
[refs] Integrating later legacy MBQL changes from the beginning

Some later changes in my local branches required a bunch more changes to
`lib.convert`, `legacy-mbql.*`, etc. that broke loads of tests and
conflicted with downstream changes. This integrates them all in one PR
for a less noisy path.

metamben reviewed (approved):
🏅

appleby reviewed (approved):
Checked out the branch, verified that idents are created for the expected clauses and persisted to the db for saved questions

bshepherdson commented on src/metabase/legacy_mbql/schema.cljc:
```diff
@@ -1466,7 +1496,17 @@
-      (empty? (set/intersection (set breakout) (set fields))))]])
+      (empty? (set/intersection (set breakout) (set fields))))]
+   ;; TODO: Bring this back?
+   #_[:fn
```
Yes. I've improved the comment. It breaks loads of tests but not in a way that adds value.

bshepherdson commented on src/metabase/legacy_mbql/schema.cljc:
```diff
@@ -1436,20 +1443,43 @@
+    [:source-table       {:optional true} SourceTable]
+    [:aggregation        {:optional true} [:sequential {:min 1} Aggregation]]
+    [:aggregation-idents {:optional true} [:ref ::AggregationIdents]]
+    [:breakout           {:optional true} [:sequential {:min 1} Field]]
```
It got missed, it's there now.

bshepherdson commented on src/metabase/lib/schema/join.cljc:
```diff
@@ -63,6 +63,7 @@
    [:stages      [:ref :metabase.lib.schema/stages]]
    [:conditions  ::conditions]
    [:alias       ::alias]
+   [:ident       :string]
```
Yes; I've switched this to `::common/non-blank-string`.

bshepherdson commented on src/metabase/query_processor/util/transformations/nest_breakouts.clj:
```diff
@@ -57,6 +57,11 @@
 
+(defn- copy-ident [to from]
+  (if-let [ident (lib.options/ident from)]
+    (lib.options/update-options to assoc :ident ident)
```
Done.

bshepherdson commented on test/metabase/test/data/mbql_query_impl.cljc:
```diff
@@ -162,12 +165,66 @@
+(defn populate-idents
+  "Internal impl fn for `data/mbql-query` macro.
+
+  Fills in `:aggregation-idents` and `:expression-idents` where missing."
```
I'm not sure what [this comment](https://github.com/metabase/metabase/pull/50085#discussion_r1866214677) means.

metamben replied:
Your message seems to be off, as does [mine](https://github.com/metabase/metabase/pull/50085#discussion_r1866214677). Strange.

bshepherdson commented on test/metabase/test/data/mbql_query_impl.cljc:
```diff
@@ -162,12 +165,66 @@
+      ;; This metadata is used by query= and similar to ignore mismatches if the mbql-query macro generated idents on
+      ;; the expectation side. It includes only the *new*, macro-generated idents.
+      true         (vary-meta assoc :generated-idents
+                              {:expression-idents  (set (keys fresh-expr-idents))
+                               :aggregation-idents (set (remove (set (keys aggregation-idents)) (keys agg-idents)))
+                               :breakout-idents    (set (remove (set (keys breakout-idents))    (keys brk-idents)))
+                               :join-idents        (set (keep-indexed (fn [i join]
```
I've moved them all to `:generated-paths/aggregation-idents` etc. and updated the `=?` handling.

bshepherdson commented on test/metabase/test_runner/assert_exprs.clj:
```diff
@@ -21,11 +24,44 @@
+  [joins original-joins generated-join-idents]
+  ;; Recursively strip generated idents off the inner queries of joins!
+  (let [joins (map strip-generated-idents joins original-joins)]
+    (if generated-join-idents
```
I think you're right. It works anyway, but that's probably because the inner operation is a no-op if `generated-join-idents` is `#{}`. Fixed.

metamben commented on src/metabase/legacy_mbql/normalize.cljc:
```diff
@@ -392,12 +392,24 @@
+                     :expressions        normalize-expressions-tokens
+                     :expression-idents  (fn [m]
+                                           (update-keys m #(if (keyword? %)
+                                                             (name %)
```
My suggestion is about gracefully handling namespaced keywords.

ranquild committed:
Fix loading in segments

bshepherdson committed:
Add e2e helper to fill in idents; use it in createQuestion et al

bshepherdson committed:
Stop enforcing :ident on joins, until we add it in the BE

bshepherdson committed:
Fix some Postgres tests that got missed
